pkg/provider/chef/chef.go

```diff
@@ -0,0 +1,314 @@
+}
+
+// GetSecret returns a databagItem present in the databag. format example: databagName/databagItemName.
+func (providerchef *Providerchef) GetSecret(ctx context.Context, ref v1beta1.ExternalSecretDataRemoteRef) ([]byte, error) {
+	if utils.IsNil(providerchef.databagService) {
+		return nil, fmt.Errorf(errUninitalizedChefProvider)
+	}
+
+	key := ref.Key
+	databagName := ""
+	databagItem := ""
+	nameSplitted := strings.Split(key, "/")
+	if len(nameSplitted) > 1 {
+		databagName = nameSplitted[0]
+		databagItem = nameSplitted[1]
+	}
+	log.Info("fetching secret value", "databag Name:", databagName, "databag Item:", databagItem)
+	if databagName != "" && databagItem != "" {
+		return getSingleDatabagItem(providerchef, databagName, databagItem, ref.Property)
+	}
+
+	return nil, fmt.Errorf(errInvalidFormat)
+}
+
+func getSingleDatabagItem(providerchef *Providerchef, dataBagName, databagItemName, propertyName string) ([]byte, error) {
+	ditem, err := providerchef.databagService.GetItem(dataBagName, databagItemName)
+	metrics.ObserveAPICall(ProviderChef, CallChefGetDataBagItem, err)
+	if err != nil {
+		return nil, fmt.Errorf(errNoDatabagItemFound, databagItemName, dataBagName)
+	}
+
+	jsonByte, err := json.Marshal(ditem)
+	if err != nil {
+		return nil, fmt.Errorf(errUnableToConvertToJSON)
+	}
+
+	if propertyName != "" {
+		return getPropertyFromDatabagItem(jsonByte, propertyName)
+	}
+
+	return jsonByte, nil
+}
+
+/*
+A path is a series of keys separated by a dot.
+A key may contain special wildcard characters '*' and '?'.
+To access an array value use the index as the key.
+To get the number of elements in an array or to access a child path, use the '#' character.
+The dot and wildcard characters can be escaped with '\'.
+
+refer https://github.com/tidwall/gjson#:~:text=JSON%20byte%20slices.-,Path%20Syntax,-Below%20is%20a
+*/
+func getPropertyFromDatabagItem(jsonByte []byte, propertyName string) ([]byte, error) {
+	result := gjson.GetBytes(jsonByte, propertyName)
+
+	if !result.Exists() {
+		return nil, fmt.Errorf(errNoDatabagItemPropertyFound, propertyName)
+	}
+	return []byte(result.Str), nil
+}
+
+// GetSecretMap returns multiple k/v pairs from the provider, for dataFrom.extract.key
+// dataFrom.extract.key only accepts dataBagName, example : dataFrom.extract.key: myDatabag
+// databagItemName or Property not expected in key.
+func (providerchef *Providerchef) GetSecretMap(ctx context.Context, ref v1beta1.ExternalSecretDataRemoteRef) (map[string][]byte, error) {
+	if utils.IsNil(providerchef.databagService) {
+		return nil, fmt.Errorf(errUninitalizedChefProvider)
+	}
+	databagName := ref.Key
+
+	if strings.Contains(databagName, "/") {
+		return nil, fmt.Errorf(errInvalidDataform)
+	}
+	getAllSecrets := make(map[string][]byte)
+	log.Info("fetching all items from databag:", databagName)
+	dataItems, err := providerchef.databagService.ListItems(databagName)
+	metrics.ObserveAPICall(ProviderChef, CallChefListDataBagItems, err)
```

Comment:
I think we would want this metric even if it failed in the first `IsNil` call, right?

Reply:
I understand that it logs the error, but this metric will no longer be available if chef is uninitialized.

Meh, maybe, this is fine. :D